dask/array/gufunc.py

```diff
@@ -0,0 +1,455 @@
+        set ``vectorize=True``). If this function returns multiple outputs,
+        ``output_core_dims`` has to be set as well.
+    *args : numeric
+        Input arrays or scalars to the callable function.
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
+    output_dtypes : dtype or list of dtypes, keyword only
+        dtype or list of output dtypes.
+    output_sizes : dict, optional, keyword only
+        Optional mapping from dimension names to sizes for outputs. Only used if
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool, keyword only
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+    allow_rechunk: Optional, bool, keyword only
+        Allows rechunking, otherwise chunk sizes need to match and core
+        dimensions are to consist only of one chunk.
+        Warning: enabling this can increase memory usage significantly.
+        Defaults to ``False``.
+
+    Returns
+    -------
+    Single dask.array.Array or tuple of dask.array.Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    >>> import numpy as np
+    >>> def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    >>> a = da.random.normal(size=(10,20,30), chunks=(5, 10, 30))
+    >>> mean, std = da.apply_gufunc(stats, "(i)->(),()", a, output_dtypes=2*(a.dtype,))
+    >>> mean.compute().shape
+    (10, 20)
+
+
+    >>> def outer_product(x, y):
+    ...     return np.einsum("i,j->ij", x, y)
+    >>> a = da.random.normal(size=(   20,30), chunks=(10, 30))
+    >>> b = da.random.normal(size=(10, 1,40), chunks=(5, 1, 40))
+    >>> c = da.apply_gufunc(outer_product, "(i),(j)->(i,j)", a, b, output_dtypes=a.dtype, vectorize=True)
+    >>> c.compute().shape
+    (10, 20, 30, 40)
+
+    References
+    ----------
+    .. [1] http://docs.scipy.org/doc/numpy/reference/ufuncs.html
+    .. [2] http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+    """
+    output_dtypes = kwargs.pop("output_dtypes", None)
+    output_sizes = kwargs.pop("output_sizes", None)
+    vectorize = kwargs.pop("vectorize", None)
+    allow_rechunk = kwargs.pop("allow_rechunk", False)
+    if output_dtypes is None:
+        raise ValueError("Must specify `output_dtypes` of output array(s)")
+
+    # Input processing:
+    ## Signature
+    if isinstance(signature, str):
+        core_input_dimss, core_output_dimss = _parse_gufunc_signature(signature)
+    else:
+        core_input_dimss, core_output_dimss = signature
+
+    ## Determine nout
+    nout = None if not isinstance(core_output_dimss, list) else len(core_output_dimss)
+
+    ## Assert output_dtypes
+    if nout is not None and not (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
```

Comment:
You should coerce output_dtypes first to a uniform form. `output_dtypes='dd'` could also be valid (at least for `np.vectorize`) to indicate two doubles.